src/spec/test/metaprogramming/ASTMatcherTestingTest.groovy

```diff
@@ -0,0 +1,123 @@
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package metaprogramming
+
+import groovy.transform.CompileDynamic
+import groovy.transform.CompileStatic
+import org.codehaus.groovy.ast.ASTNode
+import org.codehaus.groovy.ast.ClassNode
+import org.codehaus.groovy.ast.MethodNode
+import org.codehaus.groovy.ast.Parameter
+import org.codehaus.groovy.ast.expr.BinaryExpression
+import org.codehaus.groovy.ast.expr.Expression
+import org.codehaus.groovy.ast.stmt.BlockStatement
+import org.codehaus.groovy.control.CompilePhase
+import org.codehaus.groovy.control.SourceUnit
+import org.codehaus.groovy.macro.matcher.ASTMatcher
+import org.codehaus.groovy.macro.transform.MacroClass
+import org.codehaus.groovy.transform.AbstractASTTransformation
+import org.codehaus.groovy.transform.GroovyASTTransformation
+import org.codehaus.groovy.transform.GroovyASTTransformationClass
+
+import java.lang.annotation.ElementType
+import java.lang.annotation.Retention
+import java.lang.annotation.RetentionPolicy
+import java.lang.annotation.Target
+
+import static org.codehaus.groovy.ast.ClassHelper.Integer_TYPE
+import static org.codehaus.groovy.ast.tools.GeneralUtils.*
+
+/**
+ * Created by dev on 6/30/16.
+ */
+class ASTMatcherTestingTest extends GroovyTestCase {
+
+    // tag::testexpression[]
+    void testTestingSumExpression() {
+        use(ASTMatcher) {                 // <1>
+            TwiceASTTransformation sample = new TwiceASTTransformation()
+            Expression referenceNode = macro {
+                a + a                     // <2>
+            }.withConstraints {           // <3>
+                placeholder 'a'           // <4>
+            }
+
+            assert sample
+                .sumExpression
+                .matches(referenceNode)   // <5>
+        }
+    }
+    // end::testexpression[]
+
+    // tag::executiontesting[]
+    void testASTBehavior() {
+        assertScript '''
+        package metaprogramming
+
+        @Twice
+        class AAA {
+
+        }
+
+        assert new AAA().giveMeTwo(1) == 2
+        '''
+    }
+    // end::executiontesting[]
+}
+
+@Retention(RetentionPolicy.SOURCE)
+@Target([ElementType.TYPE])
+@GroovyASTTransformationClass(["metaprogramming.TwiceASTTransformation"])
+@interface Twice { }
+
+// tag::twiceasttransformation[]
+@GroovyASTTransformation(phase = CompilePhase.INSTRUCTION_SELECTION)
+class TwiceASTTransformation extends AbstractASTTransformation {
+
+    static final String VAR_X = 'x'
+
+    @Override
+    void visit(ASTNode[] nodes, SourceUnit source) {
+        ClassNode classNode = (ClassNode) nodes[1]
+        MethodNode giveMeTwo = getTemplateClass(sumExpression)
+            .methods
+            .find { it.name == 'giveMeTwo' }
```

Comment:
can be replaced with `getTemplateClass(sumExpression).getDeclaredMethods('giveMeTwo').first()`